catalog: span use-after-finish

**Describe the problem**

From https://github.com/cockroachdb/cockroach/pull/76894, saw this failure (master @ 97cc8de9483f0e123d4a5cbde0a7584fe9a89609)

```
=== RUN   TestSnapshotAfterTruncationWithUncommittedTail
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestSnapshotAfterTruncationWithUncommittedTail2372052169
    test_log_scope.go:80: use -show-logs to present logs inline
panic: use of Span after Finish. Span: prepare stmt. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 292096 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc003f1bb00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:185 +0x245
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc003f1bb00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:387 +0x2f
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:395
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x93ddb98, 0xc00a9677d0})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x4f
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth({0x93ddb98, 0xc00a9677d0}, 0x1, 0x2, 0x0, {0x625bac6, 0x22}, {0xc007597b70, 0x2, 0x2})
	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:55 +0x2b0
github.com/cockroachdb/cockroach/pkg/util/log.Warningf(...)
	github.com/cockroachdb/cockroach/bazel-out/k8-dbg/bin/pkg/util/log/log_channels_generated.go:449
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.releaseLease({0x93ddb98, 0xc00a9677d0}, 0xc001519710, 0xc0078fd500)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/lease.go:543 +0x377
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.acquireNodeLease.func1()
	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/lease.go:510 +0x4f2
github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall(0xc000639710, 0xc0072ce550, {0xc0058a7eb7, 0x9}, 0xc00aa2acc0)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:128 +0x68
created by github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).DoChan
	github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:121 +0x59b
I220222 18:42:11.234987 1 (gostd) testmain.go:1556  [-] 1  Test //pkg/kv/kvserver:kvserver_test exited with error code 2
```

This suggests a problem with the span lifecycles in the `catalog` package.


Jira issue: CRDB-13329

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -476,11 +476,8 @@ func acquireNodeLease(ctx context.Context, m *Manager, id descpb.ID) (bool, erro
 		// of the first context cancels other callers to the `acquireNodeLease()` method,
 		// because of its use of `singleflight.Group`. See issue #41780 for how this has
 		// happened.
-		baseCtx := m.ambientCtx.AnnotateCtx(context.Background())
-		// AddTags and not WithTags, so that we combine the tags with those
-		// filled by AnnotateCtx.
-		baseCtx = logtags.AddTags(baseCtx, logtags.FromContext(ctx))
-		newCtx, cancel := m.stopper.WithCancelOnQuiesce(baseCtx)
+		lt := logtags.FromContext(ctx)
+		ctx, cancel := m.stopper.WithCancelOnQuiesce(logtags.AddTags(m.ambientCtx.AnnotateCtx(context.Background()), lt))
 		defer cancel()
 		if m.isDraining() {
 			return nil, errors.New("cannot acquire lease when draining")
@@ -490,7 +487,7 @@ func acquireNodeLease(ctx context.Context, m *Manager, id descpb.ID) (bool, erro
 		if newest != nil {
 			minExpiration = newest.getExpiration()
 		}
-		desc, expiration, err := m.storage.acquire(newCtx, minExpiration, id)
+		desc, expiration, err := m.storage.acquire(ctx, minExpiration, id)
 		if err != nil {
 			return nil, err
 		}
@@ -499,7 +496,7 @@ func acquireNodeLease(ctx context.Context, m *Manager, id descpb.ID) (bool, erro
 		t.mu.takenOffline = false
 		defer t.mu.Unlock()
 		var newDescVersionState *descriptorVersionState
-		newDescVersionState, toRelease, err = t.upsertLeaseLocked(newCtx, desc, expiration)
+		newDescVersionState, toRelease, err = t.upsertLeaseLocked(ctx, desc, expiration)
 		if err != nil {
 			return nil, err
 		}
@@ -507,7 +504,7 @@ func acquireNodeLease(ctx context.Context, m *Manager, id descpb.ID) (bool, erro
 			m.names.insert(newDescVersionState)
 		}
 		if toRelease != nil {
-			releaseLease(newCtx, toRelease, m)
+			releaseLease(ctx, toRelease, m)
 		}
 		return true, nil
 	})

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -507,7 +507,7 @@ func acquireNodeLease(ctx context.Context, m *Manager, id descpb.ID) (bool, erro
 			m.names.insert(newDescVersionState)
 		}
 		if toRelease != nil {
-			releaseLease(ctx, toRelease, m)
+			releaseLease(newCtx, toRelease, m)
 		}
 		return true, nil
 	})

--- a/pkg/testutils/lint/lint_test.go
+++ b/pkg/testutils/lint/lint_test.go
@@ -2132,7 +2132,14 @@ func TestLint(t *testing.T) {
 			stream.GrepNot(`In file included from.*(start|runtime)_jemalloc\.go`),
 			stream.GrepNot(`include/jemalloc/jemalloc\.h`),
 
+			// Allow shadowing for variables named err, pErr (proto-errors in kv) and
+			// ctx. For these variables, these names are very common and having too
+			// look for new names to avoid shadowing is too onerous or even
+			// counter-productive if it makes people use the wrong variable by
+			// mistake.
 			stream.GrepNot(`declaration of "?(pE|e)rr"? shadows`),
+			stream.GrepNot(`declaration of "ctx" shadows`),
+
 			// This exception is for hash.go, which re-implements runtime.noescape
 			// for efficient hashing.
 			stream.GrepNot(`pkg/sql/colexec/colexechash/hash.go:[0-9:]+: possible misuse of unsafe.Pointer`),

